public/app/plugins/datasource/tempo/datasource.ts

```diff
@@ -453,7 +470,14 @@ export class TempoDatasource extends DataSourceWithBackend<TempoQuery, TempoJson
           grafana_version: config.buildInfo.version,
         });
 
-        const jsonData = JSON.parse(this.uploadedJson as string);
+        let uploadedString;
+        if (typeof this.uploadedJson !== 'string') {
+          console.error('Unexpected type for uploadedJson. Expected string, got', typeof this.uploadedJson);
+          uploadedString = this.uploadedJson.toString(); // tentative fallback
+        } else {
+          uploadedString = this.uploadedJson;
+        }
+        const jsonData = JSON.parse(uploadedString);
```

Comment:
`uploadedJson` is set by the `FileDropzone` component from `@grafana/ui`. We have in fact:
```
```
<FileDropzone
  onLoad={(result) => {
    this.props.datasource.uploadedJson = result;
    ...
  }}
/>
```
and the value of `result` can be `string` or `ArrayBuffer`. I tried it and I always got a `string`, but I don't see a way to ensure that.

The best solution I can think of is:
- Move the check on the type where we use `FileDropzone`; something like
```
<FileDropzone
  onLoad={(result) => {
    if (typeof result !== "string") {
       // handle the ArrayBuffer
    } else {
      this.props.datasource.uploadedJson = result;
   }
  }}
/>
```
- Change the type of `uploadedJson` to just be `string` (potentially `null`)

Do you have a better proposal?

Reply:
Basically yes. The `FileDropzone` decides how to read it based on this prop and deafaults to string. So the typing just does not reflect that this is based on the props. https://github.com/grafana/grafana/blob/7469f58709e69fba76550cbda9adbd119d7a8921/packages/grafana-ui/src/components/FileDropzone/FileDropzone.tsx#L38

So yeah in the `onLoad` it should be string because of that and we can check and throw error if it isn't string because based on the API it shouldn't be anything else. There is no compiler typesafety in that unfortunatelly but for that FileDropzone would have to be typed better.